articles/rosbags.md

```diff
@@ -0,0 +1,482 @@
+While its main purpose will be to store ROS messages, the abstraction layer shall be ROS 2 agnostic and really only deal with binary data and generic meta information.
+This allows to manually insert custom data such as [pcap](https://en.wikipedia.org/wiki/Pcap) recordings, if suitable meta information is available.
+
+### Data encoding
+
+We define the data encoding as the format in which ROS messages are represented in binary form.
+The encoding has to be uniquely described and to be linked to the binary representation in the data storage.
+That is, to be able to interpret its binary representation and thus restore the ROS message.
+At this point, the content of this binary representation is not important for the storage API.
+We specify though, that every topic has to have a single encoding and no messages with different encodings can be stored for one topic.
+
+### Data storage
+
+We introduce a separate layer of API for interfacing with the underlying data storage.
+In order to write a message, an abstract representation of a message has to be given as input.
+Such a representation comprises the binary blob for the actual messages and a key-value pair of meta data.
+The data encoding description is part of this meta data.
+While researching for "the best data storage", we came to the conclusion that there is no single best data storage which fits all needs.
+We therefore want to be able to support multiple data storage solutions (e.g. SQLite or [rosbag format2](http://wiki.ros.org/Bags/Format/2.0)).
+The main focus is thus to create a powerful enough API which a) supports multiple data storages and is b) able to transfer the data from one storage to another.
+A typical use case for this is to have a very write efficient storage format running when recording highly performance critical data (e.g. on a car). Then, in a second step post-process the recorded data to be stored in another storage format which is better suited to introspect the data.
+The API for accessing the underlying storage thus has to be generic enough to independently describe a message and being interpreted accordingly to correctly store the message for each solution.
+Similarly, a query format shall be in place which allows to fetch/filter messages from the data storage into the binary blob and its meta data.
+
+#### SQLite
+
+We chose [SQLite](https://www.sqlite.org/about.html) as the default solution for this implementation.
+Contrary to other relational database systems, it doesn't require any external SQL Server, but is self-contained.
+It is also open source, [extensively tested](https://www.sqlite.org/testing.html) and known to have a large [community](https://www.sqlite.org/support.html).
+The Gazebo team created a nicely composed [comparison](https://osrfoundation.atlassian.net/wiki/spaces/GAZ/pages/99844178/SQLite3+Proposal) between SQLite and the existing rosbag format.
```

Comment:
The link https://osrfoundation.atlassian.net/wiki/spaces/GAZ/pages/99844178/SQLite3+Proposal doesn't appear to be working :-(

Reply:
Here is a back of this document (I think): https://drive.google.com/file/d/1mmxPSv6doMSnF0iu6jCvCi_SOY6Su1AD/view?usp=sharing